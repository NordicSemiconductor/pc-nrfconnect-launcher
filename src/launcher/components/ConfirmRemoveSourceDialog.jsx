/* Copyright (c) 2015 - 2018, Nordic Semiconductor ASA
 *
 * All rights reserved.
 *
 * Use in source and binary forms, redistribution in binary form only, with
 * or without modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions in binary form, except as embedded into a Nordic
 *    Semiconductor ASA integrated circuit in a product or a software update for
 *    such product, must reproduce the above copyright notice, this list of
 *    conditions and the following disclaimer in the documentation and/or other
 *    materials provided with the distribution.
 *
 * 2. Neither the name of Nordic Semiconductor ASA nor the names of its
 *    contributors may be used to endorse or promote products derived from this
 *    software without specific prior written permission.
 *
 * 3. This software, with or without modification, must only be used with a Nordic
 *    Semiconductor ASA integrated circuit.
 *
 * 4. Any software provided in binary form under this license must not be reverse
 *    engineered, decompiled, modified and/or disassembled.
 *
 * THIS SOFTWARE IS PROVIDED BY NORDIC SEMICONDUCTOR ASA "AS IS" AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY, NONINFRINGEMENT, AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL NORDIC SEMICONDUCTOR ASA OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
 * TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import React from 'react';
import { ConfirmationDialog } from 'pc-nrfconnect-shared';
import { bool, func, string } from 'prop-types';

const ConfirmRemoveSourceDialog = ({
    isVisible,
    source,
    onConfirm,
    onCancel,
}) => (
    <ConfirmationDialog
        isVisible={isVisible}
        title="Remove app source"
        text={`Are you sure to remove "${source}" source along with any apps installed from it?`}
        okButtonText="Yes, remove"
        cancelButtonText="Cancel"
        onOk={() => onConfirm(source)}
        onCancel={onCancel}
    />
);

ConfirmRemoveSourceDialog.propTypes = {
    isVisible: bool.isRequired,
    source: string,
    onConfirm: func.isRequired,
    onCancel: func.isRequired,
};

ConfirmRemoveSourceDialog.defaultProps = {
    source: null,
};

export default ConfirmRemoveSourceDialog;
