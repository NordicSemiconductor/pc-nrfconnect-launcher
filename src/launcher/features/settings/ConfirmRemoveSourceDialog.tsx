/*
 * Copyright (c) 2015 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import React from 'react';
import { ConfirmationDialog } from 'pc-nrfconnect-shared';

import { useLauncherDispatch, useLauncherSelector } from '../../util/hooks';
import { removeSource } from './settingsEffects';
import { getSettings, hideRemoveSourceDialog } from './settingsSlice';

export default () => {
    const { removeSource: source, isRemoveSourceDialogVisible: isVisible } =
        useLauncherSelector(getSettings);
    const dispatch = useLauncherDispatch();

    return (
        <ConfirmationDialog
            isVisible={isVisible}
            title="Remove app source"
            text={`Are you sure to remove "${source}" source along with any apps installed from it?`}
            okButtonText="Yes, remove"
            cancelButtonText="Cancel"
            onOk={() => {
                dispatch(hideRemoveSourceDialog());
                dispatch(removeSource(source!)); // eslint-disable-line @typescript-eslint/no-non-null-assertion
            }}
            onCancel={() => dispatch(hideRemoveSourceDialog())}
        />
    );
};
