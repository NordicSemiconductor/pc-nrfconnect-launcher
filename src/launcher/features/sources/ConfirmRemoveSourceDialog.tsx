/*
 * Copyright (c) 2015 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import React from 'react';
import { ConfirmationDialog } from 'pc-nrfconnect-shared';

import { useLauncherDispatch, useLauncherSelector } from '../../util/hooks';
import { removeSource } from './sourcesEffects';
import {
    getIsRemoveSourceVisible,
    getSourceToRemove,
    hideRemoveSource,
} from './sourcesSlice';

export default () => {
    const dispatch = useLauncherDispatch();
    const isVisible = useLauncherSelector(getIsRemoveSourceVisible);
    const sourceToRemove = useLauncherSelector(getSourceToRemove);

    return (
        <ConfirmationDialog
            isVisible={isVisible}
            title="Remove app source"
            text={`Are you sure to remove "${sourceToRemove}" source along with any apps installed from it?`}
            okButtonText="Yes, remove"
            cancelButtonText="Cancel"
            onOk={() => {
                dispatch(hideRemoveSource());
                dispatch(removeSource(sourceToRemove!)); // eslint-disable-line @typescript-eslint/no-non-null-assertion
            }}
            onCancel={() => dispatch(hideRemoveSource())}
        />
    );
};
