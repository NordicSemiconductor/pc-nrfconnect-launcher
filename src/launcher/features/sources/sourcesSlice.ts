/*
 * Copyright (c) 2022 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import type { PayloadAction } from '@reduxjs/toolkit';
import { createSlice } from '@reduxjs/toolkit';

import {
    getDoNotRemindDeprecatedSources as getPersistedDoNotRemindDeprecatedSources,
    setDoNotRemindDeprecatedSources as setPersistedDoNotRemindDeprecatedSources,
} from '../../../common/persistedStore';
import { Source, SourceName, SourceUrl } from '../../../ipc/sources';
import type { RootState } from '../../store';

type Hidden = typeof hidden;
const hidden = { isVisible: false } as const;

type WarnAboutMissingTokenOnAddSource = {
    isVisible: true;
    sourceToAdd: SourceUrl;
};
export const isWarningAboutMissingTokenOnAddSource = (
    warning: MissingTokenWarning
): warning is WarnAboutMissingTokenOnAddSource =>
    warning.isVisible && 'sourceToAdd' in warning;

type WarnAboutMissingTokenOnMigratingSources = {
    isVisible: true;
    sourcesWithRestrictedAccessLevel: Source[];
};
export const isWarningAboutMissingTokenOnMigratingSources = (
    warning: MissingTokenWarning
): warning is WarnAboutMissingTokenOnMigratingSources =>
    warning.isVisible && 'sourcesWithRestrictedAccessLevel' in warning;

type MissingTokenWarning =
    | Hidden
    | WarnAboutMissingTokenOnAddSource
    | WarnAboutMissingTokenOnMigratingSources;

export type State = {
    sources: Source[];
    isAddSourceVisible: boolean;
    sourceToRemove: null | SourceName;
    deprecatedSources: Source[];
    doNotRemindDeprecatedSources: boolean;
    sourceToAdd: null | SourceUrl;
    missingTokenWarning: MissingTokenWarning;
};

const initialState: State = {
    sources: [],
    isAddSourceVisible: false,
    sourceToRemove: null,
    deprecatedSources: [],
    doNotRemindDeprecatedSources: getPersistedDoNotRemindDeprecatedSources(),
    sourceToAdd: null,
    missingTokenWarning: hidden,
};

const sourcesWithout = (sources: Source[], sourceNameToBeRemoved: SourceName) =>
    sources.filter(
        existingSource => existingSource.name !== sourceNameToBeRemoved
    );

const slice = createSlice({
    name: 'sources',
    initialState,
    reducers: {
        setSources(state, { payload: sources }: PayloadAction<Source[]>) {
            state.sources = [...sources];
        },
        addSource(
            state,
            {
                payload: newSource,
            }: PayloadAction<{
                name: SourceName;
                url: SourceUrl;
            }>
        ) {
            state.sources = [
                ...sourcesWithout(state.sources, newSource.name),
                newSource,
            ];
        },
        removeSource(state, { payload: name }: PayloadAction<SourceName>) {
            state.sources = sourcesWithout(state.sources, name);
        },
        showAddSource(state) {
            state.isAddSourceVisible = true;
        },
        hideAddSource(state) {
            state.isAddSourceVisible = false;
        },
        showRemoveSource(state, { payload: name }: PayloadAction<SourceName>) {
            state.sourceToRemove = name;
        },
        hideRemoveSource(state) {
            state.sourceToRemove = null;
        },
        showDeprecatedSources(
            state,
            { payload: sources }: PayloadAction<Source[]>
        ) {
            state.deprecatedSources = [...sources];
        },
        hideDeprecatedSources(state) {
            state.deprecatedSources = [];
        },
        doNotRemindDeprecatedSources(state) {
            state.doNotRemindDeprecatedSources = true;
            setPersistedDoNotRemindDeprecatedSources();
        },
        warnAddLegacySource(state, { payload: url }: PayloadAction<SourceUrl>) {
            state.sourceToAdd = url;
        },
        clearSourceToAdd(state) {
            state.sourceToAdd = null;
        },
        warnAboutMissingTokenOnAddSource(
            state,
            { payload: sourceToAdd }: PayloadAction<SourceUrl>
        ) {
            state.missingTokenWarning = {
                isVisible: true,
                sourceToAdd,
            };
        },
        warnAboutMissingTokenOnMigratingSources(
            state,
            { payload: sources }: PayloadAction<Source[]>
        ) {
            state.missingTokenWarning = {
                isVisible: true,
                sourcesWithRestrictedAccessLevel: [...sources],
            };
        },
        hideWarningAboutMissingToken(state) {
            state.missingTokenWarning = hidden;
        },
    },
});

export default slice.reducer;

export const {
    setSources,
    addSource,
    removeSource,
    showAddSource,
    hideAddSource,
    showRemoveSource,
    hideRemoveSource,
    showDeprecatedSources,
    hideDeprecatedSources,
    doNotRemindDeprecatedSources,
    warnAddLegacySource,
    clearSourceToAdd,
    warnAboutMissingTokenOnAddSource,
    warnAboutMissingTokenOnMigratingSources,
    hideWarningAboutMissingToken,
} = slice.actions;

export const getSources = (state: RootState) => state.sources.sources;
export const getIsAddSourceVisible = (state: RootState) =>
    state.sources.isAddSourceVisible;
export const getIsRemoveSourceVisible = (state: RootState) =>
    state.sources.sourceToRemove != null;
export const getSourceToRemove = (state: RootState) =>
    state.sources.sourceToRemove;
export const getDeprecatedSources = (state: RootState) =>
    state.sources.deprecatedSources;
export const getDoNotRemindDeprecatedSources = (state: RootState) =>
    state.sources.doNotRemindDeprecatedSources;
export const getSourceToAdd = (state: RootState) => state.sources.sourceToAdd;
export const getMissingTokenWarning = (state: RootState) =>
    state.sources.missingTokenWarning;
