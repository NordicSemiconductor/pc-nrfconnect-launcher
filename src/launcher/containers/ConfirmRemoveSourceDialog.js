/*
 * Copyright (c) 2015 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import { connect } from 'react-redux';

import * as SettingsActions from '../actions/settingsActions';
import ConfirmRemoveSourceDialog from '../components/ConfirmRemoveSourceDialog';

function mapStateToProps(state) {
    return {
        source: state.settings.removeSource,
        isVisible: state.settings.isRemoveSourceDialogVisible,
    };
}

function mapDispatchToProps(dispatch) {
    return {
        onConfirm: source => {
            dispatch(SettingsActions.hideRemoveSourceDialog());
            dispatch(SettingsActions.removeSource(source));
        },
        onCancel: () => dispatch(SettingsActions.hideRemoveSourceDialog()),
    };
}

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(ConfirmRemoveSourceDialog);
