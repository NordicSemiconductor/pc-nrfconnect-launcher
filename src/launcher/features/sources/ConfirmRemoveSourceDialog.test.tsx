/*
 * Copyright (c) 2022 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import React from 'react';

import render from '../../../testrenderer';
import ConfirmRemoveSourceDialog from './ConfirmRemoveSourceDialog';
import { showRemoveSource } from './sourcesSlice';

describe('ConfirmRemoveSourceDialog', () => {
    it('is initially invisible', () => {
        expect(
            render(<ConfirmRemoveSourceDialog />).baseElement
        ).toMatchSnapshot();
    });

    it('shows the new version if one becomes available ', () => {
        expect(
            render(<ConfirmRemoveSourceDialog />, [
                showRemoveSource('the source to remove'),
            ]).baseElement
        ).toMatchSnapshot();
    });
});
