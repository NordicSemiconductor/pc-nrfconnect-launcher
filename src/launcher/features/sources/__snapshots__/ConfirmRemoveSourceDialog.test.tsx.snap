// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmRemoveSourceDialog is initially invisible 1`] = `
<body>
  <div />
</body>
`;

exports[`ConfirmRemoveSourceDialog shows the new version if one becomes available  1`] = `
<body
  class="modal-open"
  style="padding-right: 0px;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="fade modal-backdrop show"
  />
  <div
    aria-modal="true"
    class="fade dialog  modal show"
    data-focus-visible-added=""
    role="dialog"
    style="display: block;"
    tabindex="-1"
  >
    <div
      class="modal-dialog modal-dialog-centered"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header"
        >
          <div
            class="tw-flex tw-flex-row"
          >
            <b>
              Remove app source
            </b>
          </div>
        </div>
        <div
          class="modal-body"
        >
          Are you sure to remove 
          the source to remove
           source along with any apps installed from it?
        </div>
        <div
          class="modal-footer"
        >
          <div
            class="tw-preflight "
          >
            <button
              class="tw-h-8 tw-px-4 tw-text-sm tw-bg-nordicBlue tw-text-white active:enabled:tw-bg-nordicBlue-700 "
              type="button"
            >
              Yes, remove
            </button>
          </div>
          <div
            class="tw-preflight "
          >
            <button
              class="tw-h-8 tw-px-4 tw-text-sm tw-border tw-border-gray-700 tw-bg-white tw-text-gray-700  active:enabled:tw-bg-gray-50 "
              type="button"
            >
              Cancel
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
